Avoid #pragma omp simd with GCC

loveshack opened the pull request:
It just uses SIMD, and not FMA, at least for x86_64.  Versions that
support that pragma do the vectorization automatically anyhow.

devinamatthews commented:
I have had a hard time getting GCC to automatically vectorize all of the kernels (although I can't remember off the top of my head which ones where problematic). Does using the `fma` function allow FMA SIMD with the pragma? Does GCC use FMA in the automatic vectorization?

loveshack commented:
I realized this had never been addressed.  I reported the behaviour somewhere in one of the issues with a long thread.  Presumably it will improve some generic routines on recent x86 which for sizes which aren't memory-bound.  I can't easily check, but assuming the behaviour is the same on POWER as Haswell et al, it will presumably make a factor of two difference on the generic GEMM used there.

fgvanzee commented:
I'm fine with this PR (I made minor tweaks). I'll merge by end of day if nobody else has objected / chimed in by then.

devinamatthews commented:
If the only problem with the status quo is FMA, then I would rather try and get gcc to generate FMA with the simd pragma than remove it.

fgvanzee commented:
@devinamatthews I'm also fine with that. Basically, I have no dog in this fight and just want to make people happy. So I'll pause for now if you want to investigate.

devinamatthews commented:
@loveshack can you point out a particular kernel with the FMA problem? Then I will tinker.

devinamatthews commented:
Here's what I've found out so far:

1) The `fma` function is actually dangerous because if hardware FMA is not available then it calls a software emulation function which will be very slow.
2) Just writing `a*b+c` produces FMA in most cases, except as @loveshack notes for gcc when using `#pragma omp simd` for some reason.
3) Explicitly adding `-ffp-contract=fast` restores FMA even with the pragma. According to the documentation this is supposed to be the default option so I don't know why it is necessary to add it. But, since it is the default adding it should be harmless.

So, adding `-ffp-contract` for gcc (and maybe clang just in case) seems like a plausible alternative.

fgvanzee commented:
@loveshack Why don't you give `-ffp-contract=fast` a try. (Adding it to `COPTFLAGS` in your configuration's `make_defs.mk` file before running `configure` should be sufficient.) If that works for you, I'm in favor of going that route instead of disabling `#pragma omp simd` altogether.

loveshack commented:
I can't remember exactly what I did and reported, but I compared the
performance of the generic C DGEMM built for haswell with and without
the pragma enabled, and then checked the generated code.  That was after
finding it was unexpectedly slow originally by a factor of two-ish.  I
didn't actually check simple reductions.

loveshack commented:
I don't understand why you prefer an extra kludge on top of one that was
inserted because of a misconception about GCC optimization.  I don't
even know if it's always appropriate versus GCC's cost models.

Another example of trusting GCC that I may or may not have reported:
When I replaced the sse3 intrinsics in the how-to-optimize-dgemm final
version with the original plain C, and added restricts, GCC 6 (?)
produced faster code than the intrinsics, at least when run on an avx2
system.

I haven't tried to remove the hand unrolling in the generic DGEMM to see
how it compares with what GCC does for different SIMD lengths, but that
would also be interesting.

By the way, the bug report for documentation of the pragma is
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91511 but duplicates an old
one that didn't appear when I looked for a previous report.

devinamatthews commented:
> Another example of trusting GCC that I may or may not have reported:
> When I replaced the sse3 intrinsics in the how-to-optimize-dgemm final
> version with the original plain C, and added restricts, GCC 6 (?)
> produced faster code than the intrinsics, at least when run on an avx2
> system.
Was the C version was compiled with AVX2 instructions enabled? If so 
then I would expect it to be faster than hand-coded SSE. I think we both 
have a number of anectodes that support at least as many disparate 
conclusions, so I suggest that we might want to start a collection of 
micro-benchmarks and/or assembly comparisons and/or whatever to gather 
some hard data.

devinamatthews commented:
Apparently it's `-std=c99` that prevents FMA and not `#pragma omp simd`. `-ffp-contract=fast` still seems like a safe, targeted solution, though.

fgvanzee commented:
@devinamatthews Interesting. I assume `-std=gnu99` would make the difference with the FMAs?

devinamatthews commented:
Yes it does, but of course it enables a bunch of other stuff too.

fgvanzee commented:
> Yes it does, but of course it enables a bunch of other stuff too.

Yeah, this particular feature is not really worth straying from C99, IMO.

devinamatthews commented:
I created #331 which adds `-ffp-contract=fast` to all reference kernel optimizations flags, and also adds `-funsafe-math-optimizations` where it was missing. Note that at some point the latter was changed to imply the former even in C99 mode, but I'm not sure exactly which version this started with. @loveshack this should make both gcc and clang emit FMAs with or without `#pragma omp simd`. I am still in favor of keeping the pragma as I'm not convinced gcc will always autovectorize like we want it to.

jeffhammond commented:
Doesn’t GCC have a -mfma flag that doesn’t nothing but enable FMA?
-- 
Jeff Hammond
jeff.science@gmail.com
http://jeffhammond.github.io/

jeffhammond commented:
IBM has had FMA for a long time and it is 2x in the implementations with
which I am familiar.

Field should test this on AMD too. I recall their 4-port FPU design
includes 2 FMA and 2 ADD pipes so you might not see as large a difference,
depending on how stuff flows through those pipes.
-- 
Jeff Hammond
jeff.science@gmail.com
http://jeffhammond.github.io/

devinamatthews commented:
@jeffhammond the current problem is that while the FMA instruction set is enabled, specifying C99 mode is disabling the mul+add to FMA transformation because of IEEE/standards compiance.

loveshack commented:
I'm confused by this, especially the reference to c99, which I
definitely didn't change when doing a comparison.  I should go back and
have another look at it.  Unfortunately exactly what I had perished in
an un-backed up corrupted Vagrant box I was using for comparing gcc
versions.

devinamatthews commented:
So, I think #331 accomplishes the same underlying goal as this, so I'll go ahead and close. Thanks for bringing this issue up @loveshack.
